fix(no_focused_tests): detect focused test loops in noFocusedTests rule

amasotti opened the pull request:
## Summary

fixes #5493 

The current `noFocusedTests` rule identifies test methods using `.only()` (and similar cases), but it doesn't detect focused tests when these are used in a test loop (e.g. `.only.each()`).

These cases however should also raise a error for the same reason as for the current `noFocusedTests` rule.

This enhancement improves test reliability by ensuring all forms of focused tests are detected.

P.s: one additional commit only fixes one tiny typo. 

## Test Plan

- Added test cases for `.only.each()` patterns in the test files:
  - `describe.only.each([["a"], ["b"]])("%s", (a) => {})`
  - `it.only.each([["a"], ["b"]])("%s", (a) => {})`
  - `test.only.each([["a"], ["b"]])("%s", (a) => {})`
- Verified that the rule correctly identifies these focused test loops
- Confirmed that the auto-fix correctly removes the `.only` part to make these tests valid
- All tests pass when running `cargo test -p biome_js_analyze suspicious::no_focused_tests`

ematipico commented on crates/biome_js_analyze/src/lint/suspicious/no_focused_tests.rs:
```diff
@@ -58,6 +72,29 @@ impl Rule for NoFocusedTests {
+        // just recognize the "only" part and not the whole expression
+        let callee_str = callee.to_string();
+        if callee_str.contains(format!(".{ONLY_KEYWORD}.each").as_str()) {
+            if let Some(static_member) = callee.as_js_static_member_expression() {
+                if let Ok(obj) = static_member.object() {
+                    if let Some(parent) = obj.as_js_static_member_expression() {
+                        if let Ok(member) = parent.member() {
+                            if member.to_string() == ONLY_KEYWORD {
+                                return Some(member.syntax().text_trimmed_range());
+                            }
+                        }
+                    }
+                }
```
These nested `if` make the code difficult to read. I suggest using a chain of `.and_then`, when eventually `.map` the `.member()`. 

Also, don't use `to_string()`, because it allocates a string, which isn't needed. Use `member.syntax().text_trimmed()` for the comparison.

amasotti replied:
@ematipico thanks for this and the other feedbacks! 

The functional approach was also the first one I tried, but I had some trouble. I do agree with you however, I am also not a fan of nested `if`s 
I'll give it another try

ematipico commented on crates/biome_js_analyze/src/lint/suspicious/no_focused_tests.rs:
```diff
@@ -66,6 +103,15 @@ impl Rule for NoFocusedTests {
                 }
+
+                // Direct "only" member in static member expression
+                if let Some(static_member) = callee.as_js_static_member_expression() {
+                    if let Ok(member) = static_member.member() {
+                        if member.to_string() == ONLY_KEYWORD {
+                            return Some(member.syntax().text_trimmed_range());
+                        }
+                    }
+                }
```
Same thing. avoid nested `if`s, and don't allocate a string.

ematipico commented on crates/biome_js_analyze/src/lint/suspicious/no_focused_tests.rs:
```diff
@@ -80,11 +126,11 @@ impl Rule for NoFocusedTests {
                 if let Some(literal) = expression.member().ok()?.as_any_js_literal_expression() {
                     if literal.as_js_string_literal_expression().is_some()
-                        && literal.to_string() == "\"only\""
+                        && literal.to_string() == format!("\"{}\"", ONLY_KEYWORD)
```
`format!`  allocates a new string. Can we avoid that?

amasotti replied:
I get the point. Do you have some convention in the project? Shall we go with simple string comparison as it was before (just don't like too much the string literal in the logic if we can use the const)

Alternatively I can think of something like

```rust
literal.to_string() == &("\"" + ONLY_KEYWORD + "\"")
```

ematipico commented on crates/biome_js_analyze/src/lint/suspicious/no_focused_tests.rs:
```diff
@@ -115,27 +161,63 @@ impl Rule for NoFocusedTests {
 
+        // Special handling for test.only.each() pattern
+        let callee_str = callee.to_string();
+        if callee_str.contains(format!(".{ONLY_KEYWORD}.each").as_str()) {
+            if let Some(static_member) = callee.as_js_static_member_expression() {
+                if let Ok(obj) = static_member.object() {
+                    if let Some(parent) = obj.as_js_static_member_expression() {
+                        if let Ok(member) = parent.member() {
+                            if member.to_string() == ONLY_KEYWORD {
```
Same thing here, let's avoid nested ifs with a more functional approach, and let's avoid the allocation of a string

ematipico commented on crates/biome_js_analyze/src/lint/suspicious/no_focused_tests.rs:
```diff
@@ -115,27 +161,63 @@ impl Rule for NoFocusedTests {
             };
+        } else if let Some(static_member) = callee.as_js_static_member_expression() {
+            // Check if this is a direct ".only" member expression
+            if let Ok(member) = static_member.member() {
+                if member.to_string() == ONLY_KEYWORD {
```
String allocation

ematipico reviewed (commented):
Thank you @amasotti for the contribution. I highlighted the places where we can improve the code

amasotti commented:
@ematipico I've changed strategy a bit here (and rebased on `main` since inbetween there were updates to the main branch).

I'll try to sum them up here:

- I've noticed that in `expr_ext.rs` (in the crate `bjome_js_syntax`) already had quite few helfpful methods for pattern discovery, many of which also written by you. I've added two new ones that detect focused tests and `only.each` patterns
- I've refactored the implementation of `no_focused_tests.rs` to use these methods

De facto, what we're checking for here is 4 well distinct patterns:

- focused tests identified by prepended `f` (e.g. `fdescribe`)
- focused tests identified by the `only` keyword as static member of the expression
- focused tests identified by `only` as computed member (e.g. `describe["only"]`)
- chained expression with `only` and `each` (the bug reported in the issue)

For this reason I found - at least imho - much more readable to assign to the main functions `run` and `action` only the "orchestrator" role and extract the logic in smaller helper functions. This should make the implementation more readable, easy to debug and easy to expand if we want to add more patterns.

Wdyt?

P.s: Theoretically there are more patterns to cover (e.g. `fdescribe.each()` or the redundant (code smell) but possible `fdescribe.only()`. I avoided including them as well, in order to keep the PR manageable.

ematipico commented on .changeset/all-loops-strive.md:
```diff
@@ -0,0 +1,6 @@
+---
+
+Enhanced no_focused_test rule to also recognize usage in loops. Solves #5493
+Implement new helper methods in `biome_js_syntax/src/expr_ext.rs` to search for focused tests with or without loops.
```
```suggestion
```

We usually document only the changes for our end users. We don't document changes to our Rust crates (for now).

ematipico reviewed (approved):
Thank you @amasotti for you contribution, the code is very good and readable! I left some feedbacks around the changeset, we can merge it once we address them :)

amasotti commented:
@ematipico rebased on `main` and [example added](e9a0a6a216ce6b238f0f1c1e19115a32c1c166e0)
